README
NAME
    Convert::TBX::Min - Convert TBX-Min to TBX-Basic

VERSION
    version 0.05

SYNOPSIS
            use Convert::TBX::Min 'min2basic';
            min2basic('/path/to/file'); # XML string pointer okay too

DESCRIPTION
    This module converts TBX-Min XML into TBX-Basic XML.

FUNCTIONS
  "min2basic"
    Converts TBX-Min data into TBX-Basic data. The input may be either a
    TBX::Min object or data to be passed to TBX::Min's "new_from_xml" in
    TBX::Min constructor. The return value is a scalar ref containing the
    TBX-Basic XML document as a UTF-8-encoded string.

SEE ALSO
    min2basic (the included script)
    TBX::Min
    Convert::TBX::Basic

    Schema for validating TBX documents, as well as more information about
    individual dialects, is available on GitHub
    <https://github.com/byutrg/TBX-Spec>.

AUTHOR
    Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Alan K. Melby.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Convert-TBX-Min

0.05      2014-02-11 16:26:17 America/Los_Angeles
    Rereleasing after having uploaded to wrong account :/
0.04      2014-02-11 15:05:33 America/Los_Angeles
    fixed failing 02-script.t on *nix
0.03      2014-02-06 11:56:46 America/Los_Angeles
    add min2basic executable script
    require newer version of TBX::Min
    update documentation
    fix copyright name
0.02      2014-01-17 19:14:55 America/Los_Angeles
    - print titleStmt before sourceDesc (required by TBX Checker)
0.01      2014-01-11 15:44:16 America/Los_Angeles
	initial release


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Convert TBX-Min to TBX-Basic\r",
  "dist_author" => [
    "Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "Convert-TBX-Min",
  "dist_version" => "0.05",
  "license" => "perl",
  "module_name" => "Convert::TBX::Min",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Exporter::Easy" => 0,
    "TBX::Min" => "0.06",
    "XML::Twig" => 0,
    "XML::Writer" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/min2basic"
  ],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "Data::Section::Simple" => 0,
    "Devel::FindPerl" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Path::Tiny" => 0,
    "Test::Base" => 0,
    "Test::LongString" => 0,
    "Test::More" => "0.88",
    "Test::NoWarnings" => 0,
    "Test::XML" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Convert TBX-Min to TBX-Basic\r",
  "AUTHOR" => "Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Convert-TBX-Min",
  "EXE_FILES" => [
    "bin/min2basic"
  ],
  "LICENSE" => "perl",
  "NAME" => "Convert::TBX::Min",
  "PREREQ_PM" => {
    "Exporter::Easy" => 0,
    "TBX::Min" => "0.06",
    "XML::Twig" => 0,
    "XML::Writer" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Data::Section::Simple" => 0,
    "Devel::FindPerl" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Path::Tiny" => 0,
    "Test::Base" => 0,
    "Test::LongString" => 0,
    "Test::More" => "0.88",
    "Test::NoWarnings" => 0,
    "Test::XML" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Convert/TBX/Min.pm
#
# This file is part of Convert-TBX-Min
#
# This software is copyright (c) 2014 by Alan K. Melby.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Convert::TBX::Min;
use strict;
use warnings;
use TBX::Min 0.06;
use XML::Writer;
use XML::Twig;
use Exporter::Easy (
	OK => ['min2basic']
);

# report parsing errors from TBX::Min in the caller's namespace,
# not ours
our @CARP_NOT = __PACKAGE__;

# mappings between TBX-Min and TBX-Basic picklists
my %status_map = (
    preferred => 'preferredTerm-admn-sts',
    admitted => 'admittedTerm-admn-sts',
    notRecommended => 'deprecatedTerm-admn-sts',
    obsolete => 'supersededTerm-admn-sts'
);

# convert input file if called as a script
min2basic(@ARGV) unless caller;

our $VERSION = '0.05'; # VERSION

# ABSTRACT: Convert TBX-Min to TBX-Basic


sub min2basic {
	my ($input) = @_;
	my $min;
	if(ref $input eq 'TBX::Min'){
		$min = $input;
	}else{
	   $min = TBX::Min->new_from_xml($input);
	}
	my $martif = XML::Twig::Elt->new(martif => {type => 'TBX-Basic'});
    $martif->set_pretty_print('indented');
	if($min->source_lang){
		$martif->set_att('xml:lang' => $min->source_lang);
	}
    my $header = _make_header($min);
    $header->paste($martif);
    _make_text($min)->paste('after' => $header);

    my $twig = XML::Twig->new();
    $twig->set_doctype('martif', 'TBXBasiccoreStructV02.dtd');
    $twig->set_encoding('UTF-8');
    $twig->set_root($martif);
	return \$twig->sprint;
}

# create the martifHeader element from the TBX::Min input
sub _make_header {
    my ($min) = @_;
    my $header = XML::Twig::Elt->new('martifHeader');
    XML::Twig::Elt->new(p => {type => 'XCSURI'}, 'TBXBasicXCSV02.xcs')->
        wrap_in('encodingDesc')->paste($header);

    my $file_desc = XML::Twig::Elt->new('fileDesc');
    $file_desc->paste($header);
    XML::Twig::Elt->new(title => $min->id)->
        wrap_in('titleStmt')->paste($file_desc);

    my $source_desc = XML::Twig::Elt->new('sourceDesc')->
        paste(last_child => $file_desc);

    my @header_atts;
    for my $header_att (qw(creator description directionality license)){
        if(my $value = $min->$header_att){
            push @header_atts, "$header_att: $value";
        }
    }
    if(@header_atts){
        for my $att(@header_atts){
            XML::Twig::Elt->new(p => $att)->paste($source_desc);
        }
    }
    # need a default source description
    if(not $min->description){
        XML::Twig::Elt->new(p => $min->id . ' (generated from UTX)')->
            paste($source_desc);
    }

    return $header;
}

# create the body element from the TBX::Min input
sub _make_text {
    my ($min) = @_;
    my $body = XML::Twig::Elt->new('body');

    for my $concept (@{$min->entries}){
        my $entry = XML::Twig::Elt->new(
            'termEntry' => {id => $concept->id})->paste(
            last_child => $body);
        if(my $subject_field = $concept->subject_field){
            XML::Twig::Elt->new(descrip => {type => 'subjectField'},
                $subject_field)->paste($entry);
        }
        for my $lang_group (@{$concept->lang_groups}){
            my $lang_el = XML::Twig::Elt->new(
                langSet => {'xml:lang' => $lang_group->code})->
                paste(last_child => $entry);
            for my $term_group (@{$lang_group->term_groups}){
                my $term_el = XML::Twig::Elt->new('tig');
                $term_el->paste(last_child => $lang_el);
                XML::Twig::Elt->new(
                    term => $term_group->term)->paste($term_el);
                if(my $status = $term_group->status){
                    XML::Twig::Elt->new(termNote =>
                        {type => 'administrativeStatus'},
                        $status_map{$status})->
                        paste(last_child => $term_el);
                }
                if(my $pos = $term_group->part_of_speech){
                    XML::Twig::Elt->new(termNote =>
                        {type => 'partOfSpeech'}, $pos)->
                        paste(last_child => $term_el);
                }
                if(my $customer = $term_group->customer){
                    XML::Twig::Elt->new(admin =>
                        {type => 'customerSubset'}, $customer)->
                        paste(last_child => $term_el);
                }
                if(my $note = $term_group->note){
                    XML::Twig::Elt->new(note => $note)->
                        paste(last_child => $term_el);
                }
            }
        }
    }
    return $body->wrap_in('text');
}

1;

__END__

=pod

=head1 NAME

Convert::TBX::Min - Convert TBX-Min to TBX-Basic

=head1 VERSION

version 0.05

=head1 SYNOPSIS

	use Convert::TBX::Min 'min2basic';
	min2basic('/path/to/file'); # XML string pointer okay too

=head1 DESCRIPTION

This module converts TBX-Min XML into TBX-Basic XML.

=head1 FUNCTIONS

=head2 C<min2basic>

Converts TBX-Min data into TBX-Basic data. The input may be either
a TBX::Min object or data to be passed to TBX::Min's
L<TBX::Min/new_from_xml> constructor. The return value is a scalar
ref containing the TBX-Basic XML document as a UTF-8-encoded string.

=head1 SEE ALSO

=over

=item L<min2basic> (the included script)

=item L<TBX::Min>

=item L<Convert::TBX::Basic>

=back

Schema for validating TBX documents, as well as more information
about individual dialects, is available on
L<GitHub|https://github.com/byutrg/TBX-Spec>.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Alan K. Melby.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
